misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt

```diff
@@ -0,0 +1,126 @@
+package misk.metrics.digester
+
+import misk.metrics.Histogram
+import misk.metrics.HistogramRegistry
+
+/** DigestMetric contains the contents of an individual metrics within a histogram */
+class DigestMetric(
+  internal val digest: SlidingWindowDigest<*>,
+  private val labels: List<String>
+) {
+  /** Adds an observation to the metric */
+  fun observe(value: Double) {
+    digest.observe(value)
+  }
+
+  /** Returns the labels for the metric */
+  fun labels(): List<String> {
+    return labels
+  }
+
+  /** Returns the SlidingWindowDigest registered for the metric */
+  fun digest(): SlidingWindowDigest<*> {
+    return digest
+  }
+}
+
+/** TDigestHistogram stores histograms which records metrics */
+class TDigestHistogram <T : TDigest<T>> constructor(
+  private val name: String,
+  private val help: String,
+  private val quantiles: List<Double>,
+  private val tDigest: () -> SlidingWindowDigest<*>
+) : Histogram {
+
+  val metrics: MutableMap<Long, DigestMetric> = mutableMapOf()
+
+  /** Records a new metric within the histogram */
+  override fun record(duration: Double, vararg labelValues: String) {
+    if (!metrics.containsKey(key(labelValues))) {
+      val metric = DigestMetric(tDigest(), labelValues.asList())
+      metrics.put(key(labelValues), metric)
+    }
+    metrics.getValue(key(labelValues)).observe(duration)
+  }
+
+  /** Returns the number of windows within the histogram */
+  override fun count(vararg labelValues: String): Int {
+    if (metrics.containsKey(key(labelValues))) {
+      return metrics.getValue(key(labelValues)).digest.windows.count()
+    }
+    return  0
+  }
+
+  /** Returns a metric of the histogram. Order of labels matters */
+  fun getMetric(vararg labelValues: String): DigestMetric? {
+    return metrics[key(labelValues)]
+  }
+
+  /** Returns the name of the histogram */
+  fun getName(): String {
+    return name
+  }
+
+  /** Returns the help of the histogram */
+  fun getHelp(): String {
+    return help
+  }
+
+  /** Returns the quantiles of the histogram */
+  fun getQuantiles(): List<Double> {
+    return quantiles
+  }
+
+  /** Generates a unique key to store each metric under */
+  fun key(labels: Array<out String>): Long {
+    var hash: Long = 0
+    labels.forEach { key ->
+      hash += key.hashCode()
+    }
+    return hash
+  }
+}
+
+/** Default quantiles that can be used to register with a new histogram */
+val defaultQuantiles = mapOf(0.5 to 0.05, 0.75 to 0.02, 0.95 to 0.01, 0.99 to 0.001, 0.999 to 0.0001)
+
+/** TDigestHistogramRegistry registers all TDigestHistograms */
+class TDigestHistogramRegistry <T : TDigest<T>> constructor(
```

Comment:
There need to be some interface tweaks to match how this will be used in practice.

Current state:
```
    val registry = TDigestHistogramRegistry<VeneurDigest>()

    val histogram = registry.newHistogram(
      name = "request_latency_ms",
      help = "hello",
      labelNames = listOf("foo", "bar"),
      quantiles = mapOf(0.75 to 999.999) // why is this a map?
    )

    // Why does get metric return a nullable?
    histogram.getMetric("x", "y")?.observe(1.23)
```

Expected usage, I think:
```
    val registry = TDigestHistogramRegistry<VeneurDigest>()

    val histogram = registry.newHistogram(
      name = "request_latency_ms",
      help = "hello",
      labelNames = listOf("foo", "bar"),
      quantiles = defaultQuantiles, // or omitted to get defaults automatically
    )

    histogram.metric("x", "y").observe(1.23)
    // or _additional_ possibility
    histogram.metric(mapOf("foo" to "x", "bar" to "y")).observe(1.23)
```

Reply:
Its a map because prometheus summaries require an error of quantiles estimation (https://prometheus.io/docs/practices/histograms/#errors-of-quantile-estimation). It looks like the prometheus go client provides the algorithm for this:

"A summary would have had no problem calculating the correct percentile value in both cases, at least if it uses an appropriate algorithm on the client side (like the one used by the Go client)."